### PR TITLE
Track new kinds of data

### LINKED ISSUE #13: Record edit time in messages table
Right now the `messages` table has an `is_edited` column, but it does not store the time that it was edited.

My proposal is to add a datetime column called `timestamp`. If a post is new, this column is identical to the value extracted from the message snowflake. If the message has been edited or deleted, this column is updated to when that event occurred.

### LINKED ISSUE #18: Track nicknames
As it is right now, the `users` table has no clean way to track nicknames. However with the creation of a new table, it is possible.

I propose a new table called `nicknames` with the given schema:

| `user_id`    | `guild_id`   | `nickname`              |
|--------------|--------------|-------------------------|
| `BigInteger` | `BigInteger` | `Optional[Unicode(32)]` |

(with foreign keys for `user_id` and `guild_id`)

This way we have a relational solution to storing nicknames despite each guild having a potentially different value.

### LINKED ISSUE #19: Track role membership
We can track who is a member of each role in a relational way.

I propose a new table, possibly called `role_membership`, with the given schema:

| `role_id` | `guild_id` | `user_id` |
|---|---|---|
| `BigInteger` | `BigInteger` | `BigInteger` |

(with foreign keys for all three columns)

### LINKED ISSUE #20: Add changed_roles column to channels
The `voice_channels` table has a `changed_roles` column (an `ARRAY(BigInteger)`) to store which roles have differing permissions for a given channel. Since this applies to normal (i.e. text) channels as well, it should also be added.

### PR DESCRIPTION
**Pull request by ammongit**
Fixes #13 
Fixes #18 
Fixes #19 
Fixes #20 

TL;DR: Some new schema things and new kinds of data being stored, like edit times.

Also I thought our ORM handling for the message ranges was unnecessarily complicated and more cumbersome to use. Now it's just a single table.

### CODE REVIEW

**Comment by ammongit** [comment]
This branch has been modified to only have the schema changes. A separate PR on the `replace-orm` branch will be made after this one is merged.

File: statbot/sql.py [hunk 1]
```diff
@@ -274,7 +307,7 @@ def _voice_channel_values(channel):
             'is_deleted': False,
             'position': channel.position,
             'bitrate': channel.bitrate,
-            'user_limit': channel.user_limit,
+            'user_limit': channel.user_limit or 0,
```
Comment (raylu-stripe) [inline statbot/sql.py]: why is the `bitrate` nullable but the `user_limit` 0?

Comment (ammongit) [reply statbot/sql.py]: Discord represents a voice channel that can hold an unlimited number of users with the value of `0`.

`bitrate` is sort of weird. The API says [it should never be `None`](https://discordpy.readthedocs.io/en/rewrite/api.html#discord.VoiceChannel.bitrate), but while running statbot on the actual server, the value returned for one the VCs was `None`.

Comment (raylu-stripe) [reply statbot/sql.py]: but if discord represents it that way, why do you need to turn (what I assume is) a `None` to a `0`?

File: statbot/sql.py [hunk 2]
```diff
@@ -689,6 +740,70 @@ def upsert_user(self, trans, user):
...
+        ups = p_insert(self.tb_nicknames) \
+                .values(values) \
+                .on_conflict_do_update(
+                        constraint='uq_nickname',
+                        set_=values,
+                )
+        trans.execute(ups)
+
+        self._delete_role_membership(trans, member)
+        self._insert_role_membership(trans, member)
```
[the remark this answers is not in the document]
Comment (raylu-stripe) [reply statbot/sql.py]: oh I misread `execute` for `commit`... where even does this get committed?

[the remark this answers is not in the document]
Comment (ammongit) [reply statbot/sql.py]: Whoever calls it does so in a `with` statement. When flow leaves that block, the transaction is automatically committed or rolled back depending on whether an exception was raised or not.

File: statbot/sql.py (hunk 1 above)
Comment (ammongit) [reply statbot/sql.py]: I... uh, I'm not sure. After I posted that last comment, I thought maybe it can't get the data because of a permissions thing or something. I will investigate this.

Comment (ammongit) [reply statbot/sql.py]: Alright I fixed this. It seems to be that the old version of discord.py was misinterpreting channel categories as voice channels for some reason.

File: statbot/util.py [hunk 3]
```diff
@@ -13,13 +13,28 @@
...
 __all__ = [
+    'member_update',
     'get_null_id',
     'get_emoji_name',
     'get_emoji_id',
     'embeds_to_json',
     'null_logger',
 ]
 
+def member_update(before, after):
```
Comment (raylu-stripe) [inline statbot/util.py]: can we not have more stuff in `util`? this entire module is an antipattern. the function is only used in `sql.py`

(also, can we call this `user_needs_update` or something?)

File: statbot/client.py [hunk 4]
```diff
@@ -322,6 +326,7 @@ def _init_sql(self, trans):
 
         with self.sql.transaction() as trans:
             self.sql.add_user(trans, member)
+            self.sql.add_member(trans, member)
```
Comment (raylu-stripe) [inline statbot/client.py]: can we just use `upsert_member` instead?

File: statbot/sql.py [hunk 5]
```diff
@@ -285,11 +318,28 @@ def _user_values(user, deleted=False):
...
+            'nickname': member.nick,
+        }
+
+    @staticmethod
+    def _role_member_values(member, role):
+        return {
+            'role_id': role.id,
+            'guild_id': role.guild.id,
+            'user_id': member.id,
+        }
```
Comment (raylu-stripe) [inline statbot/sql.py]: why do these need to be their own functions? (especially if we got rid of `add_member`)

File: statbot/util.py (hunk 3 above)
Comment (ammongit) [reply statbot/util.py]: I've renamed it to `member_needs_update`.

I don't see why this module is an antipattern, and it's not used in only `sql.py`. Like for instance the null logger is a concept that several other modules use, so I don't see why it should be duplicated in every module that uses it.

File: statbot/client.py (hunk 4 above)
Comment (ammongit) [reply statbot/client.py]: I don't see why. The event is `on_member_join`, which means that that `Member` can't already exist in the database anywhere. So just use an update.

However, I have changed the `add_user` to `upsert_user`, since it is logically possible for a user to already be tracked but not have joined a particular guild.

File: statbot/sql.py (hunk 5 above)
Comment (ammongit) [reply statbot/sql.py]: I did it because then all the object -> SQL mapping occurs in one place (per object). This way if certain extracted attributes are needed in multiple places, I don't have duplicated code that would require extra changes if the schema or API objects are ever changed.

File: statbot/client.py (hunk 4 above)
Comment (raylu-stripe) [reply statbot/client.py]: it would be less code. the number 1 cause of bugs is people writing code

Comment (ammongit) [reply statbot/client.py]: Huh? How would it be less code?
Whether you do an insert or an upsert, that's still code.

Comment (raylu-stripe) [reply statbot/client.py]: if you called `upsert_member` instead, you could get rid of the `add_member` code

File: statbot/sql.py (hunk 5 above)
Comment (raylu-stripe) [reply statbot/sql.py]: oh god I didn't realize there were waaaaaaaaaaaay more `_values` functions
